crates/trie/sparse-parallel/src/trie.rs

```diff
@@ -181,16 +225,110 @@ impl ParallelSparseTrie {
-        let _value = value;
-        let _masks = masks;
-        let _provider = provider;
-        todo!()
+        self.prefix_set.insert(path);
+        let existing = self.upper_subtrie.inner.values.insert(path, value.clone());
+        if existing.is_some() {
+            // upper trie structure unchanged, return immediately
+            return Ok(())
+        }
+
+        // Start at the root, traversing until we find either the node to update or a subtrie to
+        // update.
+        //
+        // We first traverse the upper subtrie for two levels, and moving any created nodes to a
+        // lower subtrie if necessary.
+        //
+        // We use `next` to keep track of the next node that we need to traverse to, and
+        // `inserted_nodes` to keep track of any nodes that were created during the traversal.
+        let mut new_nodes = Vec::new();
+        let mut next = Some(Nibbles::default());
+
+        // Traverse the upper subtrie to find the node to update or the subtrie to update.
+        //
+        // First we check the upper subtrie, and if we would update solely the upper subtrie, we
+        // call update_leaf on the upper subtrie.
+        //
+        // We stop when the next node to traverse would be in a lower subtrie, or if there are no
+        // more nodes to traverse.
+        while next.is_some_and(|next| next.len() < UPPER_TRIE_MAX_DEPTH) {
+            // checked this above, so unwrap is safe
+            let current = next.unwrap();
+
+            // Traverse the next node, keeping track of any changed nodes and the next step in the
+            // trie
+            match self.upper_subtrie.next_node(current, &path, &provider)? {
+                LeafUpdateStep::Continue { next_node } => {
+                    next = Some(next_node);
+                }
+                LeafUpdateStep::Complete { inserted_nodes } => {
+                    new_nodes.extend(inserted_nodes);
+                    next = None;
+                }
+                LeafUpdateStep::NoOp => {
+                    next = None;
+                }
+            }
+        }
+
+        // Move nodes from upper subtrie to lower subtries
+        for node_path in &new_nodes {
+            // Skip nodes that belong in the upper subtrie
+            if SparseSubtrieType::path_len_is_upper(node_path.len()) {
+                continue
+            }
+
+            let subtrie_idx = path_subtrie_index_unchecked(node_path);
+            let node = self.upper_subtrie.nodes.remove(node_path).unwrap();
+
+            // If it's a leaf node, extract its value before getting mutable reference to subtrie
+            let leaf_value = if let SparseNode::Leaf { key, .. } = &node {
+                let mut leaf_full_path = *node_path;
+                leaf_full_path.extend(key);
+                Some((
+                    leaf_full_path,
+                    self.upper_subtrie.inner.values.remove(&leaf_full_path).unwrap(),
```

Comment:
how often do we expect to have this branch hit? If often, I would try to avoid inserting into `self.upper_subtrie.inner.values` in the beginning of the function only to remove it here.